File: backend-java/src/main/java/com/neria/manager/webhooks/WebhooksService.java
package com.neria.manager.webhooks;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.neria.manager.common.entities.Webhook;
import com.neria.manager.common.repos.WebhookRepository;
import com.neria.manager.common.services.EncryptionService;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
public class WebhooksService {
  private final WebhookRepository repository;
  private final EncryptionService encryptionService;
  private final ObjectMapper objectMapper;
  private final HttpClient httpClient = HttpClient.newHttpClient();

  public WebhooksService(
      WebhookRepository repository, EncryptionService encryptionService, ObjectMapper objectMapper) {
    this.repository = repository;
    this.encryptionService = encryptionService;
    this.objectMapper = objectMapper;
  }

  public List<Map<String, Object>> list() {
    return repository.findAllByOrderByCreatedAtDesc().stream().map(this::sanitize).toList();
  }

  public Map<String, Object> create(CreateWebhookRequest dto) {
    Webhook webhook = new Webhook();
    webhook.setId(UUID.randomUUID().toString());
    webhook.setTenantId(dto.tenantId);
    webhook.setUrl(dto.url);
    webhook.setEvents(toJson(dto.events));
    webhook.setEncryptedSecret(dto.secret != null && !dto.secret.isBlank()
        ? encryptionService.encrypt(dto.secret)
        : null);
    webhook.setEnabled(dto.enabled != null ? dto.enabled : true);
    webhook.setCreatedAt(LocalDateTime.now());
    webhook.setUpdatedAt(LocalDateTime.now());
    return sanitize(repository.save(webhook));
  }

  public Map<String, Object> update(String id, UpdateWebhookRequest dto) {
    Webhook webhook = repository.findById(id)
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Webhook not found"));

    if (dto.tenantId != null) {
      webhook.setTenantId(dto.tenantId);
    }
    if (dto.url != null) {
      webhook.setUrl(dto.url);
    }
    if (dto.events != null) {
      webhook.setEvents(toJson(dto.events));
    }
    if (dto.enabled != null) {
      webhook.setEnabled(dto.enabled);
    }
    if (dto.secret != null) {
      webhook.setEncryptedSecret(dto.secret.isBlank() ? null : encryptionService.encrypt(dto.secret));
    }
    webhook.setUpdatedAt(LocalDateTime.now());
    return sanitize(repository.save(webhook));
  }

  public void enqueue(WebhookEventPayload payload) {
    List<Webhook> hooks = repository.findByEnabled(true);
    List<Webhook> targets = hooks.stream()
        .filter(hook -> hook.getTenantId() == null || hook.getTenantId().equals(payload.tenantId))
        .filter(hook -> {
          List<String> events = parseEvents(hook.getEvents());
          return events.contains("*") || events.contains(payload.eventType);
        })
        .toList();
    for (Webhook hook : targets) {
      deliver(hook, payload);
    }
  }

  public void deliver(Webhook webhook, WebhookEventPayload payload) {
    if (!webhook.isEnabled()) {
      return;
    }
    try {
      String body = objectMapper.writeValueAsString(payload);
      HttpRequest.Builder builder = HttpRequest.newBuilder()
          .uri(URI.create(webhook.getUrl()))
          .header("Content-Type", "application/json")
          .header("User-Agent", "ProviderManagerWebhook/1.0")
          .POST(HttpRequest.BodyPublishers.ofString(body));

      if (webhook.getEncryptedSecret() != null && !webhook.getEncryptedSecret().isBlank()) {
        String secret = encryptionService.decrypt(webhook.getEncryptedSecret());
        builder.header("x-signature", signPayload(secret, body));
      }

      HttpResponse<String> response = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
      if (response.statusCode() < 200 || response.statusCode() >= 300) {
        throw new IllegalStateException(
            "Webhook delivery failed: " + response.statusCode() + " " + response.body());
      }
    } catch (Exception ex) {
      throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, ex.getMessage());
    }
  }

  private Map<String, Object> sanitize(Webhook hook) {
    return Map.of(
        "id", hook.getId(),
        "tenantId", hook.getTenantId(),
        "url", hook.getUrl(),
        "events", parseEvents(hook.getEvents()),
        "enabled", hook.isEnabled(),
        "createdAt", hook.getCreatedAt(),
        "updatedAt", hook.getUpdatedAt());
  }

  private String signPayload(String secret, String payload) {
    try {
      Mac mac = Mac.getInstance("HmacSHA256");
      mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
      byte[] raw = mac.doFinal(payload.getBytes(StandardCharsets.UTF_8));
      return toHex(raw);
    } catch (Exception ex) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Unable to sign webhook payload");
    }
  }

  private String toHex(byte[] data) {
    StringBuilder sb = new StringBuilder();
    for (byte b : data) {
      sb.append(String.format("%02x", b));
    }
    return sb.toString();
  }

  private List<String> parseEvents(String value) {
    if (value == null || value.isBlank()) {
      return List.of();
    }
    try {
      return objectMapper.readValue(value, List.class);
    } catch (Exception ex) {
      return List.of();
    }
  }

  private String toJson(Object payload) {
    try {
      return objectMapper.writeValueAsString(payload);
    } catch (JsonProcessingException ex) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid JSON payload");
    }
  }

  public static class WebhookEventPayload {
    public String eventType;
    public String tenantId;
    public Map<String, Object> data;
    public String createdAt;
  }

  public static class CreateWebhookRequest {
    public String tenantId;
    public String url;
    public List<String> events;
    public String secret;
    public Boolean enabled;
  }

  public static class UpdateWebhookRequest {
    public String tenantId;
    public String url;
    public List<String> events;
    public String secret;
    public Boolean enabled;
  }
}

File: backend-javascript/src/webhooks/webhooks.service.ts
import { Inject, Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { createHmac } from 'crypto';
import { Repository } from 'typeorm';
import { Webhook } from '../common/entities/webhook.entity';
import { EncryptionService } from '../common/services/encryption.service';
import { WEBHOOKS_QUEUE, WebhooksQueue } from '../queues/webhooks-queue';
import { CreateWebhookDto } from './dto/create-webhook.dto';
import { UpdateWebhookDto } from './dto/update-webhook.dto';

export type WebhookEventPayload = {
  eventType: string;
  tenantId: string;
  data: Record<string, unknown>;
  createdAt: string;
};

@Injectable()
export class WebhooksService {
  constructor(
    @InjectRepository(Webhook)
    private readonly webhookRepository: Repository<Webhook>,
    private readonly encryptionService: EncryptionService,
    @Inject(WEBHOOKS_QUEUE)
    private readonly queue: WebhooksQueue | null
  ) {}

  async list() {
    const hooks = await this.webhookRepository.find({ order: { createdAt: 'DESC' } });
    return hooks.map((hook) => ({
      id: hook.id,
      tenantId: hook.tenantId,
      url: hook.url,
      events: hook.events,
      enabled: hook.enabled,
      createdAt: hook.createdAt,
      updatedAt: hook.updatedAt
    }));
  }

  async create(dto: CreateWebhookDto) {
    const webhook = this.webhookRepository.create({
      tenantId: dto.tenantId ?? null,
      url: dto.url,
      events: dto.events,
      encryptedSecret: dto.secret ? this.encryptionService.encrypt(dto.secret) : null,
      enabled: dto.enabled ?? true
    });
    const saved = await this.webhookRepository.save(webhook);
    return {
      id: saved.id,
      tenantId: saved.tenantId,
      url: saved.url,
      events: saved.events,
      enabled: saved.enabled,
      createdAt: saved.createdAt,
      updatedAt: saved.updatedAt
    };
  }

  async update(id: string, dto: UpdateWebhookDto) {
    const webhook = await this.webhookRepository.findOne({ where: { id } });
    if (!webhook) {
      throw new NotFoundException('Webhook not found');
    }
    if (dto.secret !== undefined) {
      webhook.encryptedSecret = dto.secret
        ? this.encryptionService.encrypt(dto.secret)
        : null;
    }
    Object.assign(webhook, {
      tenantId: dto.tenantId ?? webhook.tenantId,
      url: dto.url ?? webhook.url,
      events: dto.events ?? webhook.events,
      enabled: dto.enabled ?? webhook.enabled
    });

    const saved = await this.webhookRepository.save(webhook);
    return {
      id: saved.id,
      tenantId: saved.tenantId,
      url: saved.url,
      events: saved.events,
      enabled: saved.enabled,
      createdAt: saved.createdAt,
      updatedAt: saved.updatedAt
    };
  }

  async enqueue(event: WebhookEventPayload) {
    const hooks = await this.webhookRepository.find({ where: { enabled: true } });
    const targets = hooks.filter((hook) => {
      const matchesTenant = !hook.tenantId || hook.tenantId === event.tenantId;
      const matchesEvent = hook.events.includes('*') || hook.events.includes(event.eventType);
      return matchesTenant && matchesEvent;
    });

    if (!this.queue) {
      await Promise.all(targets.map((hook) => this.deliver(hook.id, event)));
      return;
    }

    const queue = this.queue;
    await Promise.all(
      targets.map((hook) =>
        queue.add('dispatch', {
          webhookId: hook.id,
          event
        })
      )
    );
  }

  async signPayload(secret: string, payload: string) {
    return createHmac('sha256', secret).update(payload).digest('hex');
  }

  async deliver(webhookId: string, event: WebhookEventPayload) {
    const webhook = await this.webhookRepository.findOne({ where: { id: webhookId } });
    if (!webhook || !webhook.enabled) {
      return { skipped: true };
    }

    const payload = JSON.stringify(event);
    const secret = webhook.encryptedSecret
      ? this.encryptionService.decrypt(webhook.encryptedSecret)
      : null;

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'User-Agent': 'ProviderManagerWebhook/1.0'
    };

    if (secret) {
      headers['x-signature'] = await this.signPayload(secret, payload);
    }

    const response = await fetch(webhook.url, {
      method: 'POST',
      headers,
      body: payload
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`Webhook delivery failed: ${response.status} ${text}`);
    }

    return { delivered: true };
  }
}
